src/main/scala/analysis/DSAUtility.scala

```diff
@@ -0,0 +1,983 @@
+          val params: Map[Procedure, Set[Variable]]
+         ) {
+
+  // DSNodes owned by this graph, only updated once analysis is done,
+  val nodes: mutable.Set[DSN] = mutable.Set()
+
+
+  // represent callees in proc
+  val callsites: mutable.Set[CallSite] = mutable.Set()
+
+  val mallocRegister = Register("R0", 64)
+  val stackPointer = Register("R31", 64)
+
+  // this is the mapping from offsets/positions on the stack to their representative DS nodes
+  val stackMapping: mutable.Map[BigInt, DSN] =
+    computeDomain(IntraProcIRCursor, Set(proc)).toSeq.sortBy(_.toShortString).foldLeft(Map[BigInt, DSN]()) {
+      (results, pos) => stackBuilder(pos, results)
+    }.to(collection.mutable.Map)
+
+
+  /**
+   * this function takes a stackMapping and updates it based on a memory load and memory store
+   * @param pos memory load or store IL position
+   * @param index memory location of load or store
+   * @param size size of the load or store
+   * @param m stack mapping
+   * @return updated stack mapping
+   */
+  private def visitStackAccess(pos: CFGPosition, index: Expr, size: Int, m: Map[BigInt, DSN]) : Map[BigInt, DSN] =
+    val byteSize = (size.toDouble / 8).ceil.toInt
+    index match
+      case BinaryExpr(op, arg1: Variable, arg2) if varToSym.contains(pos) && varToSym(pos).contains(arg1) &&
+        evaluateExpression(arg2, constProp(pos)).isDefined =>
+        var offset = evaluateExpression(arg2, constProp(pos)).get.value
+        varToSym(pos)(arg1).foldLeft(m) { // go through all the symbolic accesses tied to arg1 at pos
+          (m, sym) =>
+            sym match
+              case SymbolicAccess(accessor, StackLocation(regionIdentifier, proc, size), symOffset) => // only consider stack accesses
+                offset = offset + symOffset
+                createStackMapping(pos.toShortString, offset, m, byteSize)
+              case _ => m
+        }
+      case arg: Variable if varToSym.contains(pos) && varToSym(pos).contains(arg) =>
+        varToSym(pos)(arg).foldLeft(m) {
+          (m, sym) =>
+            sym match
+              case SymbolicAccess(accessor, StackLocation(regionIdentifier, proc, size), offset) =>
+                createStackMapping(pos.toShortString, offset, m, byteSize)
+              case _ => m
+        }
+      case _ => m
+
+  private def stackBuilder(pos: CFGPosition, m: Map[BigInt, DSN]): Map[BigInt, DSN] = {
+    pos match
+      case Assign(variable: Variable, expr: Expr, _) =>
+        expr match
+          case MemoryLoad(mem, index, endian, size) =>
+            visitStackAccess(pos, index, size, m)
+          case _ => m
+      case MemoryAssign(mem, index: Expr, value: Expr, endian, size: Int, label) =>
+        visitStackAccess(pos, index, size, m)
+      case _ => m
+
+  }
+
+  private def createStackMapping(label: String, offset: BigInt, m:  Map[BigInt, DSN], byteSize: Int) :  Map[BigInt, DSN]=
+    if m.contains(offset) then
+      assert(!m(offset).cells(0).growSize(byteSize))
```

Comment:
This assertion can fail unexpectedly, though I foolishly didn't keep track of the example where it does - I'll try to recreate it later

Reply:
I think this would fail if there are accesses of different sizes to the same stack position. This can be handled by collapsing stack positions ie (if R31 + 8 is written to using different sizes we assume the size of R31 + 8 is the maximum size of the accesses.